app_scheduler/src/app_scheduler/app_scheduler_lib.py

```diff
@@ -24,15 +31,61 @@ def __init__(self, robot_name, yaml_path, duration, update_duration):
             '/{}/start_app'.format(self.robot_name), StartApp)
         self.stop_app = rospy.ServiceProxy(
             '/{}/stop_app'.format(self.robot_name), StopApp)
+        self.app_lock = threading.Lock()
         self.job_timer = rospy.Timer(rospy.Duration(duration), self._timer_cb)
         self.update_timer = rospy.Timer(
             rospy.Duration(update_duration), self._update_timer_cb)
+        self.pub_schedules = rospy.Publisher(
+            '~app_schedules', AppScheduleEntries, queue_size=1)
         self.sub = rospy.Subscriber(
             '/{}/application/app_status'.format(self.robot_name),
             AppStatus, self._sub_cb)
+        self.srv_add_entry = rospy.Service(
+            '~add_entry', AddEntry, self._handler_add_entry)
+        self.srv_remove_entry = rospy.Service(
+            '~remove_entry', RemoveEntry, self._handler_remove_entry)
         self._load_yaml()
         self._register_apps()
 
+    def _add_entry(self, entry):
+        app = {
+                'name': entry.name,
+                'app_name': entry.app_name,
+                'app_schedule': {}
+                }
+        if entry.app_schedule.start != '':
+            app['app_schedule']['start'] = entry.app_schedule.start
+        if entry.app_schedule.stop != '':
+            app['app_schedule']['stop'] = entry.app_schedule.stop
+        rospy.loginfo(
+            'register app schedule => name: {0}, app_name: {1}'.format(
+                app['name'], app['app_name']))
+        self.app_lock.acquire()
+        self.apps.append(app)
+        self._register_app(app)
+        self.app_lock.release()
```

Comment:
I think it is better to use `with` for `lock`.
https://docs.python.org/3/library/threading.html#using-locks-conditions-and-semaphores-in-the-with-statement

Reply:
Fixed.